### ISSUE TITLE
Feature request. Create standalone copy from a RealmObject / copyFromRealm

### ISSUE THREAD

**Issue by dmytrodanylyk** [realm/realm-java#931]
If I get object reference from realm, sometimes I want to have `raw` object (with all fields filled), not a reference, since reference may become invalid. It would be great if we will have 'out of the box' method like `getRawObject`.

E.g.
```java
Dog dog = realmDatabase.getDogById(1000).getRawObject();
```

**Comment by cmelchior** [comment]
Hi @dmytrodanylyk 
Yes, we have been discussing adding something like that internally as well, but havn't decided either way yet.

**Comment by dmytrodanylyk** [comment]
@cmelchior if you count third part people +1 from me

**Comment by cmelchior** [comment]
We very much do. That is one of the reasons we are on GitHub, so third part people can easily chime in :)

**Comment by weengs** [comment]
+1 from me too.
I use EventBus and the only thing I can effectively post are "MyModelRealmProxy" classes. Although this is doable and works, Android Studio's gradle linting system counts this as an unknown class and marks it in red.
Having access to a Raw Objects might partly solve that issue. Unless there is another way I am unaware of...

**Comment by SleeplessByte** [comment]
:+1:

**Comment by njovy** [comment]
+1 from me as well. 
This feature is essential as Realm objects cannot be accessed across threads. I currently instantiate a parcelable object using AutoParcel in Android and NSDictionary in Objective c for realm objects which is not too bad but makes boilerplate codes.

**Comment by osama-lionheart** [comment]
That might solve the threading issue as @njovy pointed out but I don't think that would be a good solution, because how would you handle relationships? will we basically get the standalone object with everything that it links to? if you did that, it might be a costly operation for large object graphs, and if you didn't and only returned the object with null relationships that wouldn't be a practical solution either.

Currently the solution that we are using to handle this problem (with the help of retrofit / rx-java) is to do all processing on background thread, then switch to UI thread, and get the same object using a shared ui-realm in order to pass it to the UI code, something like:

```
public Observable<User> getUser(String id) {
    return usersApi.getUser(id)
                .doOnNext(user -> usersStorage.saveUser(user))
                .observeOn(AndroidSchedulers.mainThread())
                .flatMap(user -> usersStorage.getUser(id));
}
```

**Comment by SleeplessByte** [comment]
Actually if #878 would work, we could pass them on in bundles. Triggering a readparcelable would trigger the re-establishment of the connection and give us a valid object with relations.

@osama-lionheart Yeah I do the same thing, it's not optimal.

**Comment by JackWink** [comment]
+1

**Comment by RyanCopley** [comment]
+1

**Comment by ahornerr** [comment]
:+1:

**Comment by bmunkholm** [comment]
The +1's @JackWink @RyanCopley @ahornerr would be very helpful if you could also quickly state which issue it would solve for you. Other solutions for passing objects across threads are on it's way, so it's important for us to understand if this is just for that purpose.

**Comment by ahornerr** [comment]
@bmunkholm 
Specifically I have an object that is being fetched once and then passed between multiple threads where I no longer need Realm to update if there are changes, I just need the data the first time it is fetched. Quick code example:

````java
Realm realm = Realm.getInstance(this);
final Server server = realm.where(Server.class).equalTo("UUID", serverUUID).findFirst();
realm.close()

//Do multi-threaded operations which require the data in the Server class but don't need to subscribe to changes

//In each one of the connector.method() methods, there are several other threads which are executed, none of which would modify any data in the Server object.
connector.testServer(server, new IServerConnectionListener() {
    ...
    connector.testAuthentication(server, new IServerAuthListener(){
        ...
        //And so on and so forth
    }
});
````
My workaround for this would be to copy the data out into a non-realm object like so:

````java
Realm realm = Realm.getInstance(this);
Server realmServer = realm.where(Server.class).equalTo("UUID", serverUUID).findFirst();
Server server = new Server();
server.setHost(realmServer.getHost());
server.setApiKey(realmServer.getApiKey());
server.setUsername(realmServer.getUsername());
server.setPassword(realmServer.getPassword());
realm.close();
````

**Comment by bmunkholm** [comment]
Ok - thanks! Yeah if you really don't want to use the object with Realm again, or see updates to it, it could be useful. There are still issues though with how to do this with objects having relations.

**Comment by ahornerr** [comment]
I understand that relations could be an issue. For my use case, there are no relations inside the `Server` object, thankfully.

**Comment by bmunkholm** [comment]
Sure - the issue fo us in supporting this in general is to find a solution that makes sense and works in all situations :-(

**Comment by RyanCopley** [comment]
@bmunkholm Could you ELI5 how relations would cause an issue? Why couldn't you just pull the entire relationship tree out?

**Comment by bmunkholm** [comment]
Nothing really prevents loops in your relations. So that would have to be detected.

**Comment by cmelchior** [comment]
Loops are not really the problem, it is more the risk of references causing the entire object graph to be copied, which could potentially take a long time as well as use a lot of memory. This leaves two choices:

1) We do it anyway and let it be up to developers to avoid that situation.
2) Add some option to control if references are copied and potentially how deep.

**Comment by RyanCopley** [comment]
Yeah a max depth copy sounds like it would work from my viewpoint. Are there any other hurdles you all know of?

**Comment by JackWink** [comment]
Late to chime in, but my use-case is similar to @ahornerr's. Having a configurable max copy depth / throwing an exception if it's exceeded sounds sane to me.

**Comment by dmytrodanylyk** [comment]
> Loops are not really the problem, it is more the risk of references causing the entire object graph to be copied, which could potentially take a long time as well as use a lot of memory. This leaves two choices:

> 1) We do it anyway and let it be up to developers to avoid that situation.
> 2) Add some option to control if references are copied and potentially how deep.

I think we can specify which objects we need to copy.

E.g.

```java
class Dog {
    int age;
    Breed breed;
    Sex sex;
    String name;
}

```

```java
Dog dog = ....getRawObject(); // copy all fields
Dog dog = ....getRawObject(Breed.class); // copy name, age and Breed
Dog dog = ....getRawObject(Sex.class); // copy name, age and Sex
Dog dog = ....getRawObject(Sex.class, Breed.class); // copy name, age, Sex and Breed
```

**Comment by yakivmospan** [comment]
+1

**Comment by aercanozcan** [comment]
+1

**Comment by kanishktripathi** [comment]
Any timeline of this feature?

**Comment by bowmanb** [comment]
+1. Came here looking for something exactly like `getRawObject()`.

**Comment by ahornerr** [comment]
I've found in several cases that using/passing the Realm object's UUID and then querying for the Realm object when necessary is just fine for me.

**Comment by kanishktripathi** [comment]
@ahornerr 
Yes, but it's not just passing within threads which is a problem.
Serializing the object is also a majot issue

**Comment by an-k** [comment]
+1

I also get RealmObjects in an asynchronous job and then just want to throw them on the bus to populate my UI / List / ... whatever. [This](https://github.com/realm/realm-java/blob/master/experimental%2FpathExperiment%2Fsrc%2Fmain%2Fjava%2Fio%2Frealm%2Fpath%2FMainActivity.java#L55to60) workaround is no option for me as all values have to be extracted by hand. My current workaround is to copy every `MyRealmFoo` into its corresponding `MyFoo` for further work, but that's tedious, too.

A mthod like 
`copyFromRealm(int referenceDepth)`
sounds perfect to me. 


> Other solutions for passing objects across threads are on it's way

Could you specify those? I'm curious now.

**Comment by cmelchior** [comment]
@an-k We are currently working on implementing async queries. It sounds like that would solve your use case without having to copy into memory. You can follow progress on it here: https://github.com/realm/realm-java/pull/1214

That said, we fully understand there are other use cases where having a copy of the data in memory might be useful (even though you loose consistency guarantees) so it is coming as well.

**Comment by an-k** [comment]
@cmelchior Sounds perfect to me and thanks for the quick response! I can hardly wait for this feature.
Also it's nice to hear that the copy feature is also coming as well!

**Comment by fuwaneko** [comment]
+1
I want to filter realm dataset in subclass of Filter to use with AutoCompleteTextView, and it uses background thread for actual filtering then passes result to UI thread where it becomes invalid.

**Comment by lucamtudor** [comment]
While I'm waiting for this feature, I started using MapStruct (https://github.com/mapstruct/mapstruct) to "detach" some RealmObjects. Here's an example https://github.com/lucamtudor/Realm-Sandbox

**Comment by blork** [comment]
Any updates on this?

**Comment by zaki50** [comment]
@blork Currently no... We are discussing how to deal with this issue, but no conclusion yet.

**Comment by RyanCopley** [comment]
Is it really under discussion? Because this thread seems dead, even after a significant amount of people thinking it's a good idea and 8 months have elapsed...

**Comment by Saketme** [comment]
@RyanCopley IMO this is not a very critical missing feature. You can easily create your own method for cloning your object.

**Comment by RyanCopley** [comment]
I never said it was, but there is an obvious large set of people who believe this should be an easier process than having the copy every property.

**Comment by fuwaneko** [comment]
I agree with @RyanCopley, it might not be a critical feature, but it's important one. Also it makes natural sense: if you have "attached" instance you should have ability to detach it.

**Comment by bmunkholm** [comment]
Just FYI, we do our outmost to diligently prioritise every feature request and bug. Bug fixes and help/support goes above anything and enhancements that have no workaround goes above those that have. That's why features like this can take a while. That said, We are getting much closer to features like this now. We have been hard at work to get out Null, Async Queries, and are still putting the finishing touches on new migration API, dynamic/untyped realms and much more flexible model classes.

**Comment by limoges** [comment]
It's alright, the feature requires a bit of a reflexion on how/what should be easy to do/should be handled.

My principal use case presently is communication from iPhone to Watch and so-forth. A feature with a good strategy for these case would be welcome. For now I turn my objects into dictionaries because that's what practical for iOS, especially for iPhone to Watch communication.

A function similar like primaryKey() where I select which fields should be imported and which should not would be all I need most of the time. There might be better solutions, I most say that I have not scrubbed the API reference to see all of Realm's abilities and implementation details; 
For the most part, it just works and the inability to access Realm objects from other thread has been a gift more than a problem because it forces MVC + Observer down everyone's throat and ends up simplifying development.

**Comment by meierjan** [comment]
+1

**Comment by CDRussell** [comment]
I've just hit this use case too.

For my needs, I use Otto to broadcast events when data is ready. My activities request data asynchronously from a data source that might pull from Realm, the network (or both), and delivers the results back to the UI layer. I don't care about realm updates in this instance.

I would like the freedom for the UI to edit these objects freely without the need to begin a realm transaction (or even know realm at all as this is not a concern for the UI at all). When the UI is finished (as only that UI will know when it is finished editing) it will call a save method which might save back to realm, the network (or both). 

The ability to detach realm results from the DB seems like a natural concern of my centralised data store layer. Manually doing these for each RealmObject just seems clunky and fragile (easy to miss a new field when it is added). 

TL;DR 
+1

**Comment by Saketme** [comment]
@CDRussell I was doing something similar in my project. The solution was easy: Instead of broadcasting the data model, I'm now broadcasting its ID. Fetching objects in Realm is incredibly fast so this was extremely easy to do.

**Comment by CDRussell** [comment]
@saketme, that sounds good for dealing with the Realm objects. But for the scenario of handing network requests too, it gets a little messier.

I guess I would either have to subscribe to events firing which contain a Realm ID as well as the full objects which come from the network instead of the DB. Possible but a little messy. It still leaves the UI having to know about Realm and handle transactions which also seems messy.

I guess I could solve the network/DB problem by persisting everything into the DB first, and pulling the new ID from it and broadcasting that.

That might help, but still leaves UI knowing way too much about my DB for my liking.

Thanks for the comment.

**Comment by RonakPatel21** [comment]
+1

**Comment by AliceDiNunno** [comment]
+1

**Comment by Zhuinden** [comment]
@AliceDiNunno what's the +1 about? It was added in 0.86.0 as `realm.copyFromRealm()`
